src/libraries/Microsoft.Extensions.Options/tests/SourceGenerationTests/Baselines/NetCoreApp/Validators.g.cs

```diff
@@ -73,11 +89,19 @@ partial class SecondValidatorNoNamespace
         var baseName = (string.IsNullOrEmpty(name) ? "SecondModelNoNamespace" : name) + ".";
         var builder = new global::Microsoft.Extensions.Options.ValidateOptionsResultBuilder();
         var context = new global::System.ComponentModel.DataAnnotations.ValidationContext(options);
+        var validationResults = new global::System.Collections.Generic.List<global::System.ComponentModel.DataAnnotations.ValidationResult>();
+        var validationAttributes = new global::System.Collections.Generic.List<global::System.ComponentModel.DataAnnotations.ValidationAttribute>();
```

Comment:
Just like the attributes are allocated as statics, could we just create these lists statically as well and reuse them?

Or at the very least, the lists could be allocated with an explicit capacity since this is known by the generator.

Reply:
do you mean allocate them globally to work with all allocated generated classes? I think this can cause thread unsafe risk. we can allocate allocate them as thread static but I am not sure if it is worth it. 

I can force creating the list with the capacity. This is a good idea.

Reply:
From what I see here, the validationAttributes list is initialized with the set of static validation attribute instances. You should therefore be able to create N lists statically, holding all the combinations of validation attributes the generated code would ever need.

Reply:
I don't think we can optimize much here without making more involving changes which can make it extremely complicated. The current mechanism tries to check if any attribute is cached and return the cache. This attribute caching check by itself is allocating a new List anyway. 

https://github.com/dotnet/extensions/blob/79baf4a9be130a9e33b3fd8cec97fadfebcae2d6/src/Generators/Microsoft.Gen.OptionsValidation/Common/Emitter.cs#L212

Without optimizing the whole thing, the result will not be saving any memory allocations. I see it will really increase it. 

I'll keep what I have for this PR, and we can talk offline for the opportunities optimizing the whole process. Note, we are just allocating once per validation type which is not that costly at the end. I'll try to look at optimizing the Clean and allocation with capacity though.